airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py

```diff
@@ -0,0 +1,91 @@
+from typing import Any, Callable, Optional
+
+import anyio
+import dagger
+from asyncclick import Context, get_current_context
+from dagger.api.gen import Client, Container
+from pipelines.cli.click_decorators import LazyPassDecorator
+from pydantic import BaseModel, Field, PrivateAttr
+
+from ..singleton import Singleton
+
+
+class ClickPipelineContext(BaseModel, Singleton):
+    """
+    A replacement class for the Click context object passed to click functions.
+
+    This class is meant to serve as a singleton object that initializes and holds onto a single instance of the
+    Dagger client, which is used to create containers for running pipelines.
+    """
+
+    dockerd_service: Optional[Container] = Field(default=None)
+    _dagger_client: Optional[Client] = PrivateAttr(default=None)
+    _click_context: Callable[[], Context] = PrivateAttr(default_factory=lambda: get_current_context)
+
+    @property
+    def params(self):
+        """
+        Returns a combination of the click context object and the click context params.
+
+        This means that any arguments or options defined in the parent command will be available to the child command.
+        """
+        ctx = self._click_context()
+        click_obj = ctx.obj
+        click_params = ctx.params
+        command_name = ctx.command.name
+
+        # Error if click_obj and click_params have the same key, and not the same value
+        intersection = set(click_obj.keys()) & set(click_params.keys())
+        if intersection:
+            for key in intersection:
+                if click_obj[key] != click_params[key]:
+                    raise ValueError(
+                        f"Your command '{command_name}' has defined options/arguments with the same key as its parent, but with different values: {intersection}"
+                    )
+
+        return {**click_obj, **click_params}
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def __init__(self, **data: dict[str, Any]):
+        """
+        Initialize the ClickPipelineContext instance.
+
+        This method checks the _initialized flag for the ClickPipelineContext class in the Singleton base class.
+        If the flag is False, the initialization logic is executed and the flag is set to True.
+        If the flag is True, the initialization logic is skipped.
+
+        This ensures that the initialization logic is only executed once, even if the ClickPipelineContext instance is retrieved multiple times.
+        This can be useful if the initialization logic is expensive (e.g., it involves network requests or database queries).
+        """
+        if not Singleton._initialized[ClickPipelineContext]:
+            super().__init__(**data)
+            Singleton._initialized[ClickPipelineContext] = True
+
+    _dagger_client_lock: anyio.Lock = PrivateAttr(default_factory=anyio.Lock)
+
+    async def get_dagger_client(self, client: Optional[Client] = None, pipeline_name: Optional[str] = None) -> Client:
```

Comment:
I'd love to have a public `PipelineContext.dagger_client` attribute / property.
It will be more convenient than running `await pipeline_context.get_dagger_client()`
Could we expose an async init class method to initialize PipelineContext where the `_dagger_client` attribute would be set? 
```python
pipeline_context = await PipelineContext.init(data)
pipeline_context.dagger_client
```
The benefit of `get_dagger_client` is that you can pass custom pipeline name though.
Let me know your opinion about this.

Reply:
Closing the loop:

I left a reply in slack on this but TLDR I dont think the gains of avoiding the `async` call here are worth it.


The Reply:
> So If we cant await in the constructor
> 
> Then we would have to explicitly call
> `pipeline_context = await PipelineContext.init(data)
> `
> somewhere in the command hierarchy
> 
> which has a few downsides (as far as im understanding it)
> 
> - extra boilerplate
> - an extra thing to remember to run
> - that extra thing may be called well before the area of the code your working on (i.e. in the group and your working on a subcommand) leaving you unaware of its existence or necessity 
> - and an extra piece to consider when chaining together subcommands
> 
> The upside is we dont use await
> 
> Im not sure thats worth the speed trade off
> 
> Though Im not sure the speed price were paying